# Introduction

[Instrumentation ticket](https://phabricator.wikimedia.org/T294777)  |  [QA ticket](https://phabricator.wikimedia.org/T297921)

# Instrumentation note
Web team has deployed the instrumentation to measure read depth of anonymous user on mobile web talk page. 
The related events will be stored in `event.mediawiki_reading_depth` schema. Sample rate is 0.1% on English Wikipedia. 



# Table of Contents



In [1]:
shhh <- function(expr) suppressPackageStartupMessages(suppressWarnings(suppressMessages(expr)))
shhh({
library(tidyverse); 
library(lubridate); 
library(scales);
library(magrittr); 
library(dplyr);
})

In [103]:
library(IRdisplay)

display_html(
'<script>  
code_show=true; 
function code_toggle() {
  if (code_show){
    $(\'div.input\').hide();
  } else {
    $(\'div.input\').show();
  }
  code_show = !code_show
}  
$( document ).ready(code_toggle);
</script>
  <form action="javascript:code_toggle()">
    <input type="submit" value="Click here to toggle on/off the raw code.">
 </form>'
)

In [99]:
options(repr.plot.width = 15, repr.plot.height = 10)

__QA on 12/16/2021__

# Check daily events

In [3]:
query <- 
"
SELECT TO_DATE(dt),  year, month,day, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY TO_DATE(dt),year, month,day
"

In [4]:
df <-  wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [5]:
df

X_c0       year month day events sessions
1   2021-11-02 2021 12     3       2     1   
2   2021-11-11 2021 12     2       2     1   
3   2021-11-11 2021 12    10       2     1   
4   2021-11-17 2021 11    23       2     1   
5   2021-11-17 2021 11    27       1     1   
6   2021-11-17 2021 12     4       4     1   
7   2021-11-18 2021 12     3      17     1   
8   2021-11-19 2021 11    25       2     1   
9   2021-11-19 2021 11    29       4     2   
10  2021-11-19 2021 12     2       1     1   
11  2021-11-20 2021 11    24       2     1   
12  2021-11-20 2021 12     1       3     1   
13  2021-11-20 2021 12    13       2     1   
14  2021-11-21 2021 11    23       2     1   
15  2021-11-21 2021 11    27       3     1   
16  2021-11-22 2021 11    30       2     1   
17  2021-11-23 2021 11    25       6     5   
18  2021-11-23 2021 11    29       2     1   
19  2021-11-23 2021 12    10       2     1   
20  2021-11-24 2021 11    24  252846 61831   
21  2021-11-24 2021 11    28       6     4   
22  2021-11-24 2021 12     5       1     1   
23  2021-11-24 2021 12    13       1     1   
24  2021-11-25 2021 11    27       4     3   
25  2021-11-26 2021 11    26  251677 64301   
26  2021-11-26 2021 12     3      10     2   
27  2021-11-27 2021 11    29      65     4   
28  2021-11-27 2021 12     2       2     1   
29  2021-11-27 2021 12     6       2     1   
30  2021-11-28 2021 11    24       1     1   
⋮   ⋮          ⋮    ⋮     ⋮   ⋮      ⋮       
261 2021-12-04 2021 12     6       5     3   
262 2021-12-04 2021 12    10       1     1   
263 2021-12-05 2021 12     5  302679 74680   
264 2021-12-06 2021 12     8       3     2   
265 2021-12-06 2021 12    12       2     1   
266 2021-12-06 2021 12    16       2     1   
267 2021-12-07 2021 12     7  312276 74212   
268 2021-12-07 2021 12    15       2     1   
269 2021-12-08 2021 12     6      60     2   
270 2021-12-08 2021 12    10      11     4   
271 2021-12-09 2021 12     5       6     1   
272 2021-12-09 2021 12     9  321458 73965   
273 2021-12-10 2021 12     8       2     1   
274 2021-12-10 2021 12    12       8     4   
275 2021-12-11 2021 12     3       1     1   
276 2021-12-11 2021 12    11  280390 70652   
277 2021-12-11 2021 12    15       4     3   
278 2021-12-12 2021 12    10      49     2   
279 2021-12-12 2021 12    14       2     2   
280 2021-12-13 2021 12     9       4     2   
281 2021-12-13 2021 12    13  314546 75724   
282 2021-12-14 2021 12    12       1     1   
283 2021-12-14 2021 12    16       1     1   
284 2021-12-15 2021 12    15  300833 71457   
285 2021-12-16 2021 12    14      15     2   
286 2021-12-18 2021 12    12       3     2   
287 2021-12-18 2021 12    16       3     2   
288 2021-12-19 2021 12    15       2     1   
289 2021-12-24 2021 11    25       2     1   
290 2022-01-02 2021 12     5       2     1

__Note:__  

The data in dt field doesn't match with the partition year, month day. It stored the date as early as 1986. It also stored future dates, in 2022, etc.



QAed on 12-20-2021

__Reason:__ 

In modern eventloggin platform dt has switched to meaning the time according to the client. So if someone, say, has the time on their phone set to 2008 or 2022, dt would reflect that. On the other hand, meta.dt (which is used to set the partition fields) is the time our server received the event, which would still be 2021. [ticket](https://phabricator.wikimedia.org/T292586#7581979)

Work-around: query data using partitions or meta.dt instead of dt field

In [14]:
query <- 
"
SELECT to_date(meta.dt) AS date_time, year, month,day, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY to_date(meta.dt), year, month,day
ORDER BY year, month,day
LIMIT 100000
"

In [15]:
df <-  wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [16]:
df

date_time  year month day events sessions
1  2021-11-23 2021 11    23   67851 16819   
2  2021-11-24 2021 11    24  253034 61902   
3  2021-11-25 2021 11    25  256473 64688   
4  2021-11-26 2021 11    26  252048 64375   
5  2021-11-27 2021 11    27  283098 70795   
6  2021-11-28 2021 11    28  305501 74499   
7  2021-11-29 2021 11    29  301952 74162   
8  2021-11-30 2021 11    30  309143 75005   
9  2021-12-01 2021 12     1  305327 72637   
10 2021-12-02 2021 12     2  295856 72881   
11 2021-12-03 2021 12     3  287434 70782   
12 2021-12-04 2021 12     4  277437 69973   
13 2021-12-05 2021 12     5  303053 74778   
14 2021-12-06 2021 12     6  321160 75728   
15 2021-12-07 2021 12     7  312676 74312   
16 2021-12-08 2021 12     8  318724 74299   
17 2021-12-09 2021 12     9  321843 74066   
18 2021-12-10 2021 12    10  292020 72921   
19 2021-12-11 2021 12    11  280674 70717   
20 2021-12-12 2021 12    12  306898 75511   
21 2021-12-13 2021 12    13  315016 75817   
22 2021-12-14 2021 12    14  296164 72306   
23 2021-12-15 2021 12    15  301287 71549   
24 2021-12-16 2021 12    16  291742 70610   
25 2021-12-17 2021 12    17  272602 68248   
26 2021-12-18 2021 12    18  280014 67660   
27 2021-12-19 2021 12    19  296141 72523   
28 2021-12-20 2021 12    20  291871 70618   
29 2021-12-21 2021 12    21  277117 68517   
30 2021-12-22 2021 12    22  282257 68079   
31 2021-12-23 2021 12    23  278053 67343   
32 2021-12-24 2021 12    24  263182 68399   
33 2021-12-25 2021 12    25  288347 73221   
34 2021-12-26 2021 12    26  310079 75155   
35 2021-12-27 2021 12    27  318410 75643   
36 2021-12-28 2021 12    28  312843 74224   
37 2021-12-29 2021 12    29  315159 75276   
38 2021-12-30 2021 12    30  302522 73586   
39 2021-12-31 2021 12    31  295294 73211

__Note:__  
Events are avaiable sicne 2021-11-23

# By wiki

In [17]:
query <- 
"
SELECT meta.domain, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY meta.domain
ORDER BY meta.domain
LIMIT 100000
"

In [19]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [20]:
df

domain                               events   sessions
1  en.jinzhao.wiki                           245      18 
2  en.turkcewiki.org                           4       2 
3  en.wiki.hancel.org                       1197       8 
4  en.wikidark.org                             2       1 
5  en.wikipedia.ahau.cf                       10       2 
6  en.wikipedia.ahmu.cf                     2934       1 
7  en.wikipedia.ahut.cf                        8       1 
8  en.wikipedia.hfut.cf                        6       2 
9  en.wikipedia.iwiki.eu.org                  14       6 
10 en.wikipedia.iwiki.uk                      43       9 
11 en.wikipedia.org                     11235815 2511379 
12 en.wikipedia.xn--3js309f.xn--kpry57d       24       4

__Note:__  
Only enabled on English wikipedia

# By namespace

In [21]:

query <- 
"
SELECT page_namespace, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY page_namespace
ORDER BY page_namespace
LIMIT 100000
"


In [22]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [23]:
df

page_namespace events   sessions
1   -1              109583   22327 
2    0            10910210 2499386 
3    1               15535    6217 
4    2                7708    1352 
5    3                3959     449 
6    4               19451    5376 
7    5                 487     115 
8    6               30640   11024 
9    7                 121      61 
10   8                 290       6 
11   9                  16       2 
12  10                4353    1184 
13  11                 143      51 
14  12                6076    2257 
15  13                  71      34 
16  14              123296   11288 
17  15                 107      31 
18 100                6888    2273 
19 101                  21       4 
20 118                1259     140 
21 119                  31      10 
22 710                  26      15 
23 711                   1       1 
24 828                  27      15 
25 829                   3       1

# BY user type: anonymous users and logged-in users

In [26]:
query <- 
"
SELECT is_anon, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY is_anon
LIMIT 100000
"


In [27]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [28]:
df


is_anon events   sessions
1 false     150123    6825 
2 true    11090179 2505231

# By action

In [29]:

query <- 
"
SELECT action, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY action
LIMIT 100000
"


In [30]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [31]:
df

action       events  sessions
1 pageLoaded   7230343 2507202 
2 pageUnloaded 4009959  961392

__Note:__  

Two types of actions are recorded: pageLoaded, pageUnloaded

# By page_length

In [32]:
query <- 
"
SELECT page_length, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY page_length
ORDER BY page_length
LIMIT 100000
"


In [33]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [34]:
df

page_length events  sessions
1       -1       10568   5367  
2        0      149593  35760  
3        1          20      6  
4        2          12      2  
5        4          19      2  
6        5          13      5  
7        6           8      3  
8        7           2      1  
9        8          16      1  
10       9          10      5  
11      10         368     73  
12      20        1558    417  
13      30        2759    670  
14      40        2366    573  
15      50        2348    599  
16      60        2387    519  
17      70        4139    694  
18      80        3860    724  
19      90        4392    799  
20     100       29387   4258  
21     200       47937   6837  
22     300       27007   5342  
23     400       15410   4392  
24     500       11652   4044  
25     600        8705   3358  
26     700        8240   3322  
27     800        8371   3542  
28     900        7881   3328  
29    1000       59816  24061  
30    2000      157009  64448  
31    3000      227476  94814  
32    4000      400822 161555  
33    5000      289731 126241  
34    6000      298311 132029  
35    7000      296538 132742  
36    8000      284444 129191  
37    9000      275657 126007  
38   10000     1335260 501486  
39   20000     1659397 609068  
40   30000     1143206 453910  
41   40000      837833 346560  
42   50000      637049 273413  
43   60000      497653 220030  
44   70000      380637 170966  
45   80000      319597 146654  
46   90000      269073 124742  
47  100000      860848 341647  
48  200000      526574 216451  
49  300000      115592  51552  
50  400000       18037   8428  
51  500000         575    224  
52  600000          25     12  
53  700000          21      5  
54  800000          11      4  
55  900000          40      4  
56 1000000          40      9  
57 2000000           2      1

# By access method

In [37]:
query <- 
"
SELECT access_method, COUNT(1) AS events, 
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY access_method
LIMIT 100000
"


In [38]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [39]:
df

access_method events  sessions
1 desktop       4387438  570453 
2 mobile web    6852864 1940983

__Note:__

Two types of access methods are recorded: desktop, mobile web

# By agent type

In [42]:
query <- "
SELECT CASE WHEN user_agent_map['device_family']='Spider' THEN 'Spider' ELSE 'User' END AS agent_type,
COUNT(1) AS events,
COUNT(DISTINCT session_token) AS sessions
FROM event.mediawiki_reading_depth
WHERE year=2021
GROUP BY CASE WHEN user_agent_map['device_family']='Spider' THEN 'Spider' ELSE 'User' END
LIMIT 100000
"



In [43]:
df <- wmfdata::query_hive(query)

Don't forget to authenticate with Kerberos using kinit



In [44]:
df

agent_type events   sessions
1 Spider          141     114 
2 User       11240161 2511334

__Note:__

We can identify and exclude spider in analysis

# Read length Distribution

In [58]:
query_total_length <- 
"
SELECT MAX(total_length) AS max, MIN(total_length) AS min, Avg(total_length) AS avg, STDDEV(total_length) AS stdev, COUNT(1) AS count
FROM event.mediawiki_reading_depth
WHERE year=2021 AND total_length is not NULL
"


In [59]:
df <- wmfdata::query_hive(query_total_length)

Don't forget to authenticate with Kerberos using kinit



In [60]:
df

max        min avg     stdev    count  
1 2603264579 0   1178408 16303164 4009959

__Note:__

For all events in 2021, the average reading length is 1178408 MS

In [76]:
query_total_length <- 
"
SELECT total_length, session_token
FROM event.mediawiki_reading_depth
WHERE year=2021 
AND total_length is not NULL
"


In [77]:
df <- wmfdata::query_hive(query_total_length)

Don't forget to authenticate with Kerberos using kinit



In [118]:
# Represent it

p <- df %>%
  ggplot( mapping=aes(x=total_length)) +
    geom_histogram(bins=30 ) +
    scale_x_log10() +
    labs ( title = "Check distribution of total read length",
         fill="") +
    theme_light(base_size=18)


In [119]:
ggsave("tmp.png",
    plot = p, width = 60, height = 30, units = "cm", dpi = "screen");

Warning message:
“Transformation introduced infinite values in continuous x-axis”
Warning message:
“Removed 2 rows containing non-finite values (stat_bin).”
